### DISCUSSION TITLE
Xamarin and CSLA

### DISCUSSION THREAD

**Discussion by CROprogrammer** [MarimerLLC/csla#3348] [Questions] [answered]
Hi,
I've written application server which I run on localhost. I also have Xamarin application which I would like to use data portal from my application server, but I'm not sure how to configure my Xamarin app to use my app server. Is there any example with CSLA 6.2.2? I'm not quite sure how to configure it

**Comment by rockfordlhotka** [comment]
It isn't really a CSLA issue, it is an issue with an Android emulator accessing localhost.

I've never figured out how to do it myself - I just host my service in Azure so my emulator can access the endpoint.

**Comment by clodewyks** [comment]
[Conveyor](https://marketplace.visualstudio.com/items?itemName=vs-publisher-1448185.ConveyorbyKeyoti) is a great tool for this, it exposes a URL that can be used to test from remote devices on the same network. Once the application is running look for the snap in which will contain the localhost url, and a LAN accessible one.

**Comment by TheCakeMonster** [comment]
If it's to do with the localhost address, then I can think of two other tools that might help as well:

Visual Studio Dev Tunnels (new in VS 17.5)
Ngrok

Both of these tools allow you to expose a local service using a public address.

If you use Visual Studio then the integration into the development environment makes Dev Tunnels a slightly simpler choice.

**Comment by CROprogrammer** [comment]
I've managed to expose AppServer's URL and I'm able to connect to it via Android Emulator (I get response when getting /api/dataportal - database connection is also valid).
But now I'm getting "Unable to read beyond the end of the stream." exception. I'm using CSLA 6.2.2 and Xamarin.
This exception does not occur when using AppServer & MVC or AppServer & WPF, only AppServer & Xamarin.

Similar problem was described in https://github.com/MarimerLLC/csla/discussions/3325, but I'm not able to solve it the way it was mentioned in the issue,

**Comment by CROprogrammer** [comment]
The problem lies in `DataPortalProxy` class in `public virtual async Task<DataPortalResult> Fetch(Type objectType, object criteria, DataPortalContext context, bool isSync)` method.


More precisely in this snippet. For some reason `serialized ` cannot be deserialized after `CallDataPortalServer` call.

```
baseCriteriaRequest.CriteriaData = SerializationFormatterFactory.GetFormatter(ApplicationContext).Serialize(criteria);
baseCriteriaRequest = ConvertRequest(baseCriteriaRequest);
byte[] serialized = SerializationFormatterFactory.GetFormatter(ApplicationContext).Serialize(baseCriteriaRequest);
serialized = await CallDataPortalServer(serialized, "fetch", GetRoutingToken(objectType), isSync);
DataPortalResponse response = (DataPortalResponse)SerializationFormatterFactory.GetFormatter(ApplicationContext).Deserialize(serialized);
response = ConvertResponse(response);
```

`CallDataPortalServer` method looks like this:
```
protected override async Task<byte[]> CallDataPortalServer(byte[] serialized, string operation, string routingToken, bool isSync)
{
  return (!isSync) ? (await CallViaHttpClient(serialized, operation, routingToken)) : CallViaWebClient(serialized, operation, routingToken);
}
```

and in my case `CallViaHttpClient` is being called.

```
private async Task<byte[]> CallViaHttpClient(byte[] serialized, string operation, string routingToken)
    {
      HttpClient httpClient = GetHttpClient();
      HttpRequestMessage httpRequestMessage = new HttpRequestMessage(HttpMethod.Post, base.DataPortalUrl + "?operation=" + CreateOperationTag(operation, ApplicationContext.VersionRoutingTag, routingToken));
      SetHttpRequestHeaders(httpRequestMessage);
      if (Options.UseTextSerialization)
      {
        httpRequestMessage.Content = new StringContent(Convert.ToBase64String(serialized));
      }
      else
      {
        httpRequestMessage.Content = new ByteArrayContent(serialized);
      }

      HttpResponseMessage httpResponse = await httpClient.SendAsync(httpRequestMessage);
      await VerifyResponseSuccess(httpResponse);
      serialized = ((!Options.UseTextSerialization) ? (await httpResponse.Content.ReadAsByteArrayAsync()) : Convert.FromBase64String(await httpResponse.Content.ReadAsStringAsync()));
      return serialized;
    }
```

`CallViaHttpClient` returns byte array, and does not produce exception.

Any idea why the response of `CallDataPortalServer` cannot be deserialized?

**Reply by TheCakeMonster** [reply]
A failure to deserialize can happen if the type you were trying to deserialize (or something nested within it) doesn't exist on the client. However, the error message doesn't sound like that is the issue; it sounds more like an attempt to read from a stream that has not been reset to the start.

Could a streaming issue perhaps be as a result of something intercepting the request on the client and attempting to write it to a log - a piece of code added to the pipeline somehow?

**Reply by rockfordlhotka** [reply]
Please make sure that the client and server are NOT configured to use text-based serialization.

Someone in a different thread had this issue because they'd copied some configuration for the client or server from a Blazor example, and in earlier versions of Blazor it was necessary to use text-based serialization. That limitation no longer exists (thankfully).

This would be in the client data portal configuration, or in the controller's constructor on the server.

**Reply by CROprogrammer** [reply]
They are not configured to use text-based serialization which makes it weird.

**Comment by CROprogrammer** [comment, marked as answer]
I solved this problem by running both my database and my app server in docker, so I could access it locally through Android emulator. The other thing was to use http instead of https, and to use text serialization in my Xamarin app and App server `UseTextSerialization = true`. This was good enough for my needs.
